File: python/noisy_gt_publisher_node.py
#!/usr/bin/env python3
"""Standalone noisy ground-truth publisher node.

Publishes noisy odometry-style GT pose/path on dedicated topics, separated
from gmmslam_node to avoid heavy-node coupling.
"""

import numpy as np
import rospy
from geometry_msgs.msg import PoseStamped
from nav_msgs.msg import Path
from scipy.spatial.transform import Rotation


def pose_to_pose_stamped(T: np.ndarray, stamp, frame_id: str) -> PoseStamped:
    ps = PoseStamped()
    ps.header.stamp = stamp
    ps.header.frame_id = frame_id
    ps.pose.position.x = float(T[0, 3])
    ps.pose.position.y = float(T[1, 3])
    ps.pose.position.z = float(T[2, 3])
    q = Rotation.from_matrix(T[:3, :3]).as_quat()
    ps.pose.orientation.x = float(q[0])
    ps.pose.orientation.y = float(q[1])
    ps.pose.orientation.z = float(q[2])
    ps.pose.orientation.w = float(q[3])
    return ps


class NoisyGTPublisherNode:
    def __init__(self):
        rospy.init_node("noisy_gt_publisher_node", anonymous=False)

        self.gt_topic = rospy.get_param("~gt_topic", "/m500_1/mavros/local_position/pose")
        self.odom_frame = rospy.get_param("~odom_frame", "world")
        self.pose_topic = rospy.get_param("~pose_topic", "/gmmslam_node/noisy_gt_pose")
        self.path_topic = rospy.get_param("~path_topic", "/gmmslam_node/noisy_gt_path")
        self.gt_init_wait_s = float(rospy.get_param("~gt_init_wait_s", 3.0))
        self.gt_noise_sigma_t = float(rospy.get_param("~gt_noise_sigma_t", 0.0))
        self.gt_noise_sigma_r = float(rospy.get_param("~gt_noise_sigma_r", 0.0))
        self.gt_noise_seed = int(rospy.get_param("~gt_noise_seed", -1))

        self._rng = (
            np.random.default_rng(self.gt_noise_seed)
            if self.gt_noise_seed >= 0
            else np.random.default_rng()
        )
        self._latest_gt_pose_raw = None
        self._gt_origin_inv = None
        self._gt_init_start_time = rospy.Time.now()
        self._last_gt_odom = None
        self._noisy_pose = np.eye(4, dtype=np.float64)
        self._noisy_path = Path()
        self._noisy_path.header.frame_id = self.odom_frame

        self.noisy_pose_pub = rospy.Publisher(self.pose_topic, PoseStamped, queue_size=1)
        self.noisy_path_pub = rospy.Publisher(self.path_topic, Path, queue_size=1)
        rospy.Subscriber(self.gt_topic, PoseStamped, self._gt_callback, queue_size=1)

        rospy.loginfo(
            f"[noisy_gt_pub] ready | gt={self.gt_topic} | pose={self.pose_topic} | path={self.path_topic}"
        )

    @staticmethod
    def _stamp_to_sec(stamp) -> float:
        return float(stamp.secs) + 1e-9 * float(stamp.nsecs)

    @staticmethod
    def _pose_msg_to_matrix(pose_msg) -> np.ndarray:
        q = np.array(
            [
                pose_msg.orientation.x,
                pose_msg.orientation.y,
                pose_msg.orientation.z,
                pose_msg.orientation.w,
            ],
            dtype=np.float64,
        )
        q_norm = np.linalg.norm(q)
        if q_norm < 1e-12:
            raise ValueError("invalid quaternion norm")
        q /= q_norm
        T = np.eye(4, dtype=np.float64)
        T[:3, :3] = Rotation.from_quat(q).as_matrix()
        T[0, 3] = float(pose_msg.position.x)
        T[1, 3] = float(pose_msg.position.y)
        T[2, 3] = float(pose_msg.position.z)
        return T

    def _ensure_origin_initialized(self, stamp) -> bool:
        if self._gt_origin_inv is not None:
            return True
        if self._latest_gt_pose_raw is None:
            return False
        now_t = self._stamp_to_sec(stamp)
        t0 = self._stamp_to_sec(self._gt_init_start_time)
        if (now_t - t0) < self.gt_init_wait_s:
            return False
        try:
            T0 = self._pose_msg_to_matrix(self._latest_gt_pose_raw.pose)
            self._gt_origin_inv = np.linalg.inv(T0)
            self._last_gt_odom = None
            self._noisy_pose = np.eye(4, dtype=np.float64)
            self._noisy_path = Path()
            self._noisy_path.header.frame_id = self.odom_frame
            rospy.loginfo("[noisy_gt_pub] GT origin initialized and noisy odometry reset")
            return True
        except Exception as e:
            rospy.logwarn_throttle(2.0, f"[noisy_gt_pub] failed to init GT origin: {e}")
            return False

    def _gt_callback(self, msg: PoseStamped):
        self._latest_gt_pose_raw = msg
        if not self._ensure_origin_initialized(msg.header.stamp):
            return
        try:
            T_gt_webots = self._pose_msg_to_matrix(msg.pose)
            T_gt_odom = self._gt_origin_inv @ T_gt_webots
        except Exception as e:
            rospy.logwarn_throttle(2.0, f"[noisy_gt_pub] invalid GT pose: {e}")
            return

        if self._last_gt_odom is None:
            self._last_gt_odom = T_gt_odom
            ps0 = pose_to_pose_stamped(self._noisy_pose, msg.header.stamp, self.odom_frame)
            self._noisy_path.header.stamp = msg.header.stamp
            self._noisy_path.poses.append(ps0)
            self.noisy_pose_pub.publish(ps0)
            self.noisy_path_pub.publish(self._noisy_path)
            return

        T_rel = np.linalg.inv(self._last_gt_odom) @ T_gt_odom
        self._last_gt_odom = T_gt_odom

        rot_noise_vec = self._rng.normal(0.0, self.gt_noise_sigma_r, size=3)
        trans_noise = self._rng.normal(0.0, self.gt_noise_sigma_t, size=3)
        T_noisy_rel = np.eye(4, dtype=np.float64)
        T_noisy_rel[:3, :3] = Rotation.from_rotvec(rot_noise_vec).as_matrix() @ T_rel[:3, :3]
        T_noisy_rel[:3, 3] = T_rel[:3, 3] + trans_noise

        self._noisy_pose = self._noisy_pose @ T_noisy_rel
        ps = pose_to_pose_stamped(self._noisy_pose, msg.header.stamp, self.odom_frame)
        self._noisy_path.header.stamp = msg.header.stamp
        self._noisy_path.poses.append(ps)
        self.noisy_pose_pub.publish(ps)
        self.noisy_path_pub.publish(self._noisy_path)


def main():
    NoisyGTPublisherNode()
    rospy.spin()


if __name__ == "__main__":
    main()

File: src/noisy_gt_publisher_node.cpp
#include <ros/ros.h>
#include <geometry_msgs/PoseStamped.h>
#include <nav_msgs/Path.h>

#include <Eigen/Core>
#include <Eigen/Geometry>
#include <cmath>
#include <random>
#include <string>

class NoisyGTPublisherNode {
public:
    NoisyGTPublisherNode(ros::NodeHandle& nh, ros::NodeHandle& pnh)
        : gt_init_start_time_(ros::Time::now()) {
        std::string gt_topic, pose_topic, path_topic;
        pnh.param<std::string>("gt_topic", gt_topic,
                               "/m500_1/mavros/local_position/pose");
        pnh.param<std::string>("odom_frame", odom_frame_, "world");
        pnh.param<std::string>("pose_topic", pose_topic,
                               "/gmmslam_node/noisy_gt_pose");
        pnh.param<std::string>("path_topic", path_topic,
                               "/gmmslam_node/noisy_gt_path");
        pnh.param<double>("gt_init_wait_s", gt_init_wait_s_, 3.0);
        pnh.param<double>("gt_noise_sigma_t", gt_noise_sigma_t_, 0.0);
        pnh.param<double>("gt_noise_sigma_r", gt_noise_sigma_r_, 0.0);
        int seed = -1;
        pnh.param<int>("gt_noise_seed", seed, -1);

        if (seed >= 0) {
            rng_.seed(static_cast<uint64_t>(seed));
        }

        noisy_path_.header.frame_id = odom_frame_;

        pose_pub_ = nh.advertise<geometry_msgs::PoseStamped>(pose_topic, 1);
        path_pub_ = nh.advertise<nav_msgs::Path>(path_topic, 1);
        gt_sub_ = nh.subscribe(gt_topic, 1,
                               &NoisyGTPublisherNode::gtCallback, this);

        ROS_INFO("[noisy_gt_pub] ready | gt=%s | pose=%s | path=%s",
                 gt_topic.c_str(), pose_topic.c_str(), path_topic.c_str());
    }

private:
    static Eigen::Matrix4d poseMsgToMatrix(const geometry_msgs::Pose& pose) {
        Eigen::Quaterniond q(pose.orientation.w, pose.orientation.x,
                             pose.orientation.y, pose.orientation.z);
        const double qn = q.norm();
        if (qn < 1e-12) {
            throw std::runtime_error("invalid quaternion (near-zero norm)");
        }
        q.normalize();
        Eigen::Matrix4d T = Eigen::Matrix4d::Identity();
        T.block<3, 3>(0, 0) = q.toRotationMatrix();
        T(0, 3) = pose.position.x;
        T(1, 3) = pose.position.y;
        T(2, 3) = pose.position.z;
        return T;
    }

    geometry_msgs::PoseStamped matrixToPoseStamped(
        const Eigen::Matrix4d& T, const ros::Time& stamp) const {
        geometry_msgs::PoseStamped ps;
        ps.header.stamp = stamp;
        ps.header.frame_id = odom_frame_;
        ps.pose.position.x = T(0, 3);
        ps.pose.position.y = T(1, 3);
        ps.pose.position.z = T(2, 3);
        Eigen::Quaterniond q(T.block<3, 3>(0, 0));
        q.normalize();
        ps.pose.orientation.x = q.x();
        ps.pose.orientation.y = q.y();
        ps.pose.orientation.z = q.z();
        ps.pose.orientation.w = q.w();
        return ps;
    }

    bool ensureOriginInitialized(const ros::Time& stamp) {
        if (gt_origin_initialized_) return true;
        if (!has_gt_pose_) return false;

        const double now_sec = stamp.toSec();
        const double t0_sec = gt_init_start_time_.toSec();
        if ((now_sec - t0_sec) < gt_init_wait_s_) return false;

        try {
            const Eigen::Matrix4d T0 = poseMsgToMatrix(latest_gt_raw_->pose);
            gt_origin_inv_ = T0.inverse();
            gt_origin_initialized_ = true;
            noisy_path_ = nav_msgs::Path();
            noisy_path_.header.frame_id = odom_frame_;
            ROS_INFO("[noisy_gt_pub] GT origin initialized");
            return true;
        } catch (const std::exception& e) {
            ROS_WARN_THROTTLE(2.0, "[noisy_gt_pub] failed to init GT origin: %s",
                              e.what());
            return false;
        }
    }

    void gtCallback(const geometry_msgs::PoseStamped::ConstPtr& msg) {
        latest_gt_raw_ = msg;
        has_gt_pose_ = true;
        if (!ensureOriginInitialized(msg->header.stamp)) return;

        Eigen::Matrix4d T_gt;
        try {
            T_gt = poseMsgToMatrix(msg->pose);
        } catch (...) {
            ROS_WARN_THROTTLE(2.0, "[noisy_gt_pub] invalid GT pose");
            return;
        }
        const Eigen::Matrix4d T_odom = gt_origin_inv_ * T_gt;

        const Eigen::Vector3d rot_noise(normal_dist_(rng_) * gt_noise_sigma_r_,
                                        normal_dist_(rng_) * gt_noise_sigma_r_,
                                        normal_dist_(rng_) * gt_noise_sigma_r_);
        const Eigen::Vector3d trans_noise(
            normal_dist_(rng_) * gt_noise_sigma_t_,
            normal_dist_(rng_) * gt_noise_sigma_t_,
            normal_dist_(rng_) * gt_noise_sigma_t_);

        const double angle = rot_noise.norm();
        Eigen::Matrix3d R_noise = Eigen::Matrix3d::Identity();
        if (angle > 1e-12) {
            R_noise =
                Eigen::AngleAxisd(angle, rot_noise.normalized()).toRotationMatrix();
        }

        Eigen::Matrix4d T_noisy = Eigen::Matrix4d::Identity();
        T_noisy.block<3, 3>(0, 0) = R_noise * T_odom.block<3, 3>(0, 0);
        T_noisy.block<3, 1>(0, 3) = T_odom.block<3, 1>(0, 3) + trans_noise;

        const auto ps = matrixToPoseStamped(T_noisy, msg->header.stamp);
        noisy_path_.header.stamp = msg->header.stamp;
        noisy_path_.poses.push_back(ps);
        pose_pub_.publish(ps);
        path_pub_.publish(noisy_path_);
    }

    ros::Subscriber gt_sub_;
    ros::Publisher pose_pub_;
    ros::Publisher path_pub_;
    std::string odom_frame_;
    double gt_init_wait_s_ = 3.0;
    double gt_noise_sigma_t_ = 0.0;
    double gt_noise_sigma_r_ = 0.0;
    std::mt19937_64 rng_{std::random_device{}()};
    std::normal_distribution<double> normal_dist_{0.0, 1.0};

    geometry_msgs::PoseStamped::ConstPtr latest_gt_raw_;
    bool has_gt_pose_ = false;
    bool gt_origin_initialized_ = false;
    Eigen::Matrix4d gt_origin_inv_ = Eigen::Matrix4d::Identity();
    ros::Time gt_init_start_time_;
    nav_msgs::Path noisy_path_;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "noisy_gt_publisher_node");
    ros::NodeHandle nh;
    ros::NodeHandle pnh("~");
    NoisyGTPublisherNode node(nh, pnh);
    ros::spin();
    return 0;
}
